### PR TITLE
Allow to override decimal separator

### PR DESCRIPTION
**Pull request by WinterComes**
Based on [#15](https://github.com/scrapinghub/price-parser/issues/15)

Add decimal_separator_hint to customize number extraction.

### CODE REVIEW

**Comment by WinterComes** [comment]
Currently WIP, I want to add more tests

File: price_parser/parser.py [hunk 1]
```diff
@@ -25,7 +25,8 @@ def amount_float(self) -> Optional[float]:
 
     @classmethod
     def fromstring(cls, price: Optional[str],
-                   currency_hint: Optional[str] = None) -> 'Price':
+                   currency_hint: Optional[str] = None,
+                   decimal_separator_hint: Optional[str] = None) -> 'Price':
```
Comment (kmike) [inline price_parser/parser.py]: What do you think about renaming it to `decimal_separator`? The difference with currency_hint is that currency_hint is just a hint and can be ignored, while here it is a hard override.

**Comment by kmike** [comment]
Looks good, thanks @WinterComes!

File: price_parser/parser.py [hunk 2]
```diff
@@ -262,13 +267,20 @@ def parse_number(num: str) -> Optional[Decimal]:
...
+    >>> parse_number("140.000", decimal_separator_hint=".")
+    Decimal('140.000')
     >>> parse_number("")
     >>> parse_number("foo")
     """
     if not num:
         return None
     num = num.strip().replace(' ', '')
-    decimal_separator = get_decimal_separator(num)
+    decimal_separator = (
```
Comment (manycoding) [inline price_parser/parser.py]: decimal_separator = decimal_separator or get_decimal_separator(num)

File: price_parser/parser.py (hunk 1 above)
Comment (WinterComes) [reply price_parser/parser.py]: yeah, it makes sense

**Comment by manycoding** [comment]
To be more clear, I think we need to handle this special case with euro.

**Comment by WinterComes** [comment]
@manycoding I agree with you. I'm looking for possible solutions. 
One of this: I think we must keep current logic with 2 decimal if we have space after €, and pick all numbers if space is missed.

File: price_parser/parser.py (hunk 2 above)
Comment (WinterComes) [reply price_parser/parser.py]: I made some changes of [regexp](https://github.com/scrapinghub/price-parser/pull/18/files#diff-1bcf8c415357b577173d2e003a172525R191)  in `extract_price_text` and added few [tests](https://github.com/scrapinghub/price-parser/pull/18/files#diff-091825c38112ff1dafc081caf2916d27R1965) which cover this. It should fix 1 case with euro.

File: tests/test_price_parsing.py [hunk 3]
```diff
@@ -1960,6 +1962,23 @@ def __eq__(self, other):
...
+PRICE_PARSING_DECIMAL_SEPARATOR_EXAMPLES = [
+    Example(None, '1250€ 600',
+            '€', '1250', 1250, "€"),
+    Example(None, '1250€ 60',
+            '€', '1250€60', 1250.60, "€"),
+    Example(None, '1250€600',
+            '€', '1250€600', 1250.600, "€"),
+]
+
+PRICE_PARSING_DECIMAL_SEPARATOR_EXAMPLES_XFAIL = [
```
Comment (manycoding) [inline tests/test_price_parsing.py]: I don't think we need XFAIL at all, as I see the fails show either that it produces unexpected outcome with wrong input or it produces wrong outcome with valid input (corner cases).

Comment (WinterComes) [reply tests/test_price_parsing.py]: Thanks for a quick review, I removed this one.

File: price_parser/parser.py [hunk 4]
```diff
@@ -182,16 +186,16 @@ def extract_price_text(price: str) -> Optional[str]:
     """
     if price.count('€') == 1:
         m = re.search(r"""
-        [\d\s.,]*?\d    # number, probably with thousand separators
-        \s*?€\s*?        # euro, probably separated by whitespace
-        \d\d
-        (?:$|[^\d])    # something which is not a digit
+        [\d\s.,]*?\d    # number, probably with thousand separators                
+        \s*?€(\s*?)?    # euro, probably separated by whitespace
```
Comment (Gallaecio) [inline price_parser/parser.py]: It looks to me like you mean to capture the optional space after the € here in capture group 1, and then below you mean to match different content depending on whether this capture group matched content or not. But I’m not aware if this feature of Python regular expressions (`(?(N)A|B)` syntax).

To me it seems like `(\s*?)?\d(?(1)\d|\d*?)` is functionally equivalent to `\s*?\d\d*?`, which goes from only accepting 2 digits after €-as-a-decimal-separator to accepting 1 or more.

Comment (WinterComes) [reply price_parser/parser.py]: In this step we don't know about of € is a decimal-separator or not, we are trying to extract price text. So, if we find a € in text, we are looking for whitespaces after  €, if we find one or more than we should pick it if we have only 2 digits, multiple digits othervise.
I'm not sure that ```(\s*?)?\d(?(1)\d|\d*?)``` is equivalent to ```\s*?\d\d*?```, because tests are not passed after replacing it.
I found a (?(id/name)yes-pattern|no-pattern) - from official re python lib documentation.

Comment (Gallaecio) [reply price_parser/parser.py]: I had no idea about `(?(id/name)yes-pattern|no-pattern)`, great to know :smiley:

**Comment by manycoding** [comment]
@kmike Can we merge?

**Comment by WinterComes** [comment]
Hi all. Is there anything else that needs to be fixed before the merge?

**Comment by kmike** [comment]
Thanks @WinterComes for a careful implementation, and @manycoding and @Gallaecio for comprehensive reviews!
